src/rgw/rgw_common.h

```diff
@@ -1074,6 +1074,16 @@ struct RGWBucketInfo {
 
   bool empty_sync_policy() const;
 
+  rgw::BucketLayout& get_layout() { return layout; }
+  const rgw::BucketLayout& get_layout() const { return layout; }
```

Comment:
these are unused. is there a reason to prefer `bucket_info.get_layout()` over `bucket_info.layout`?

Reply:
First, the many instances of `info.layout.logs.back()` in our codebase indicates a problem to me. Detailed knowledge that shouldn't need to be understood broadly is spread far and wide. If someone has an `RGWBucketInfo` object, there are some queries they should be able to make without having to understand the hierarchy of data structures underneath, and that ultimately there's a `vector` way down there that could be empty. C++ has ways to prevent this but we need to be willing to use them.

It's technical debt from the get-go.

So I think this is a place where we might want to add some syntactic vinegar to encourage devs to move those common queries into simple functions on RGWBucketInfo. And having those functions is part of that syntactic vinegar. The next piece would be making the layout variable private or protected. So now rather than writing code that calls those functions they would be encouraged to think where does this bit of detail need to be placed.

Reply:
Syntactic vinegar is like syntactic sugar, but tastes unpleasant?

Matt

Reply:
thanks Eric, this really does deserve more discussion. i 100% agree that we're still lacking abstractions around index and log layouts, and that syntax like `bucket->get_info().layout.current_index.layout.normal.num_shards` is not where we want to be. i just don't agree that new layout interfaces belong in `RGWBucketInfo`

the types in `rgw_common.h` like `RGWBucketInfo`/ `RGWBucketEntryPoint`/`RGWUserInfo` are essentially just raw structs that map directly to their encoded format in rados. they capture the 'persistent state' of our metadata objects

i don't think encapsulation at this level is a good idea; it's important that we're able to write admin tooling that can inspect and change any of these fields. and because they're in `rgw_common.h`, we don't want the types to be coupled (via member functions) to all of their associated logic

the new layout types in `rgw_bucket_layout.h` are also raw structs that match their encodings, except they add some new customization points for `BucketIndexType` and `BucketLogType` that are implemented through nesting. for example, when `current_index.layout.type == BucketIndexType::Normal`, then its `num_shards` field can be accessed under `current_index.layout.normal.num_shards`. for any other `BucketIndexType`, we should avoid touching `current_index.layout.normal` at all

the fact that `num_shards` is not valid for some index types makes an interface like `RGWBucketInfo::get_num_shards()` problematic, because our only good option is to make it assert on the index type - and this makes the caller responsible for checking the index type first. the current compromise is that `rgw_bucket_layout.h` provides a specific, unchecked `num_shards(bucket_index_normal_layout)` function, along with checked overloads for the generic layouts that assert on index type

using a `std::variant` to represent these layout customizations would do a much better job of making it impossible to access fields without checking the type. we just chose the current structure because it allowed us to quickly replace all occurences of `bucket_info.num_shards` with `bucket_info.layout.current_index.layout.normal.num_shards`

my ultimate vision here is that generic code shouldn't ever care which `BucketIndexType` or `BucketLogType` is in the `RGWBucketInfo`. it should just be able to pass its `bucket_info.layout.current_index` into a class like `RGWSI_BucketIndex_RADOS`, which itself will check the index type and act appropriately. for example:
```c++
int RGWSI_BucketIndex_RADOS::read_stats(const RGWBucketInfo& bucket_info,
                                        const bucket_index_layout_generation& index,
                                        RGWBucketEnt *stats)
{
  switch (index.layout.type) {
    case rgw::BucketIndexType::Normal:
      return read_stats_normal(bucket_info, index.layout.normal, stats);
    case rgw::BucketIndexType::Indexless:
      *stats = RGWBucketEnt{}; // indexless buckets have no stats
      return 0;
    default:
      return -ENOTSUP; // and log an error message
  }
}
```
this way, nothing needs to reach through several levels of nesting like `bucket_info.layout.current_index.layout.normal.num_shards`. a switch to `std::variant` might be a good way to 'force' this kind of refactoring

only multisite code should care about the list of `layout.logs`, and only resharding code should ever care about the `layout.resharding` state or `layout.target_index`. but resharding code can use any of the bucket index operations on either `layout.current_index` or `layout.target_index`

Reply:
Thanks for the thoughtful reply, @cbodley . I'm glad that your'e giving thought to all this. RGWBucketInfo has that raw nature, but it's also widely used in the code to represent a bucket. It already has an interface, although slim. And I think information whether a bucket is indexless and what it's "index" is should be readily available. The versioning status already is.

Certainly Zipper is doing some much needed refactoring. And there may be some room for further refactoring.